### ISSUE TITLE
mockCurrentUser does not work if a context function is used for createGraphQLHandler

### ISSUE THREAD

**Issue by wongyouth** [redwoodjs/redwood#3367]
The use case is I get client info from JWT token and need to store it in the context. Instead to provide a `getCurrentUser` function for `createGraphQLHandler`, I have to provide a custom `context` function.

// graphql.js
```
export const handler = createGraphQLHandler({
  context: getContext,
  schema,
  ...
})
```

// auth.js
```
export const getContext = async ({ event }: ContextFuncParam) => {
  if (event.headers?.authorization) {
    const { token } = await parseAuthorizationHeader(event)

    return buildContext(token)
  }
}

const buildContext = async (tokenText: string) => {
  const token = jwt.verifyToken(tokenText as string)
  const currentUser = await loadDbUser(token.sub)
  return {
    clientInfo: token.clientInfo,
    currentUser,
  }
}
```

And then all `mockCurrentUser` does not work in testing. I mean it can not mock  current user any more, but it works with Redwood version 0.35.2

I don't know if Redwood is supporting context function officially since it's not mentioned in the document,
but it will be great to be officially supported in some way.

Thanks

**Comment by dthyresson** [comment]
Hi @wongyouth I'm not sure this will help, but wanted to let you know that you do have access to the Authorization/Bearer token in your `getCurrentUser` implementation:

```js
/**
 * getCurrentUser returns the user information together with
 * an optional collection of roles used by requireAuth() to check
 * if the user is authenticated or has role-based access
 *
 * @param decoded - The decoded access token containing user info and JWT claims like `sub`
 * @param { token, SupportedAuthTypes type } - The access token itself as well as the auth provider type
 * @param { APIGatewayEvent event, Context context } - An object which contains information from the invoker
 * such as headers and cookies, and the context information about the invocation such as IP Address
 *
 * @see https://github.com/redwoodjs/redwood/tree/main/packages/auth for examples
 */
export const getCurrentUser = async (
  decoded,
  { _token, _type },
  { _event, _context }
) => {
  return { ...decoded, roles: parseJWT({ decoded }).roles }
}
```

The `_token` here is that.

But, it sounds like you want to add it to the Global Context, but could you describe why?

The reason I ask is that you do have access to the Lambda event and context in every service as well:


```js
export const post = async (
  { id }: Prisma.PostWhereUniqueInput,
  { context }
) => {
  logger.debug({ context }, '>> in Get Post - context')
  logger.debug({ headers: context.event.headers }, '>> in Get Post - headers')

  const { token } = await parseAuthorizationHeader(context.event)

  logger.debug({ token }, '>> in Get Post - JWT ')

  return db.post.findUnique({
    where: { id },
  })
}
```

Now, you can access the Bearer token via the headers:

```
api | DEBUG [2021-09-09 13:26:31.035 +0000]: >> in Get Post - headers
api |     headers: {
api |       "host": "localhost:8911",
api |       "accept": "*/*",
api |       "authorization": "Bearer asdsdasd",
api |       "accept-language": "en-us",
api |       "accept-encoding": "gzip, deflate",
api |       "content-type": "application/json",
api |       "origin": "http://localhost:8911",
api |       "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Safari/605.1.15",
api |       "connection": "keep-alive",
api |       "referer": "http://localhost:8911/graphql",
api |       "auth-provider": "custom",
api |       "content-length": "108"
api |     }
api | DEBUG [2021-09-09 13:26:31.035 +0000]: >> in Get Post - JWT 
api |     token: "asdsdasd"
a
```

Note: `parseAuthorizationHeader` assumes the auth header is present and valid:

```
api | ERROR [2021-09-09 13:28:08.342 +0000] (graphql-server): The `Authorization` header is not valid.
api |     error: {
api |       "message": "The `Authorization` header is not valid.",
```

Will occurs if there isn't one.

TLDR; I am not sure where you need the token, but you can access it from the services since the context with the headers is always passed.

**Comment by wongyouth** [comment]
Thanks @dthyresson,

The reason I store data into the global context is that I want to prepare all info in global context when a request come in and use it when needed, so I don't need to parse token in each service. I think there should be a way to store some data besides currentUser. And in my case, I want to the currentUser be the data loaded from DB.

**Comment by dthyresson** [comment]
> And in my case, I want to the currentUser be the data loaded from DB.

Also, you are away that you can load info from the database in `getCurrentUser` if you want and then augment the result ... and that return value becomes the currentUser info.

**Comment by wongyouth** [comment]
Thanks @dthyresson, If the globalContext cannot be customized, I have to put `clientInfo` into the `currentUser`. But I think it's a hack, because I load the currentUser from DB, and I don't wan't it to be polluted.
And yes, I used a custom auth implementation.

**Comment by dthyresson** [comment]
Also: https://github.com/redwoodjs/redwood/pull/3369
